### DISCUSSION TITLE
Flight Controller Design and Hardware Deployment (Project 217)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#48] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project #217](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/edit/main/projects/Flight%20Controller%20Design%20and%20Hardware%20Deployment)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by Ben0126** [comment]
Hi, I am a student at National Taitung University and interested in working and collaborating on this project

**Reply by PaulZeroOne** [reply]
Hey Ben,
I am an undergraduate @ Thapar University and I am looking forward to collaborating on this. Can you drop your email so I can contact you further?

**Reply by vmanoj1996** [reply]
Hi Ben and Paul,
Glad to see that you are interested in this project. You can find the instructions for setting up PX4 with Simulink [here](https://www.mathworks.com/help/supportpkg/px4/setup-and-configuration.html). Additionally [these examples](https://www.mathworks.com/help/supportpkg/px4/modeling-px4.html) can help you get started.

Please feel free to post any questions or thoughts you may have below.

**Comment by anchitbhuhan** [comment]
Hi, I am a graduate from Dayananda Sagar Academy Of Technology And Management (DSATM), Bangalore and interested in working on this project.

**Reply by vmanoj1996** [reply]
Glad to see your interest, Anchit. If you have not done already, please sign up for the project [here](https://forms.office.com/Pages/ResponsePage.aspx?id=ETrdmUhDaESb3eUHKx3B5lOTzSa_A6lPqq2LJKzvpM5UMTBZRkc4UTRETjFERVRDWllQRE40OUFSQS4u). 

We have shipped a Simulink quad-copter plant along with a reference controller in R2022a for performing Hardware In Loop simulations. Feel free to check [this example](https://in.mathworks.com/help/supportpkg/px4/ref/hitl-simulink-plant-example.html) and [other examples](https://in.mathworks.com/help/supportpkg/px4/modeling-px4.html) to get started.

**Comment by oh-hey-its-Jay** [comment]
Hi, my name is Jasmine an undergraduate senior from CSULA, I am currently working on the flight controller design for our project. We are currently in the process of uploading our control system to our autopilot. We are facing a problem within our project where our motors are not continuously spinning. Instead, they are stuttering when we send a thrust command through our RC. However, the appropriate motors are spinning when we give pitch, roll, and thrust commands which indicates to us that our control system is working appropriately. We are running the hardware simulation via I/O connection while having this issue, so we tried to run the control system via external mode, and we got a sample time error which stated that we need to set the sample time of all the PX4 blocks to 1ms. When we changed the sampling time in the PX4 blocks, we still got the same error message. This tells me that it may be a PWM signal rate or Code generation problem, and we have debugged the system and did not find issues with the control system or RC. We also tried the MathWorks model shown [here](https://www.mathworks.com/help/supportpkg/px4/ref/pwm-getting-started-px4.html?searchHighlight=pwm%20px4&s_tid=srchtitle_pwm%20px4_1), and the drone's motors would still stutter. I have a video of the issue attached and a picture of the error message we are receiving while running the simulation via external mode. I also have our hardware settings attached.
![Hardware setting](https://user-images.githubusercontent.com/102106899/164339287-fe8e94a3-310c-4e55-811d-e592a6a45b74.jpg)

https://user-images.githubusercontent.com/102106899/164339936-8b69a940-5fc7-4c38-b353-228027dfd81f.mov

![SampleTime error message](https://user-images.githubusercontent.com/102106899/164340106-138bb5cd-02f5-4550-ba08-2f48c46ff380.png)
![solver hardware setting](https://user-images.githubusercontent.com/102106899/164340269-be104750-407a-4c2e-92ef-bf6729b9b5b7.jpg)
![periodic](https://user-images.githubusercontent.com/102106899/164340343-7ea3d530-77d8-47bc-b8cd-a759270cd06a.jpg)

**Reply by vmanoj1996** [reply]
Hi Jasmine,
Can you confirm whether the drone motor stuttered when you deployed (build deploy start or monitor and tune) the [example](https://in.mathworks.com/help/supportpkg/px4/ref/pwm-getting-started-px4.html) model? I want to understand whether the stutter occurred during monitor and tune mode or during connected io mode. Connected IO mode is quite slow and suitable for checking sensors or actuators. For real time control, use monitor and tune or build, deploy, start option. 

About the sample time configuration, we usually set sample time to 10ms for most blocks in controller model. 1ms may be too low.

Can you share the model (along with MATLAB version) that throws "The base rate of the Simulink model 0 is...." error. It is likely due to an incorrect model configuration.

About the stutter, some debugging steps are below,
- Use a logic analyzer (or an oscilloscope) to see if PWM's on time (pulsewidth) is as expected and does not drop to <=1000us. This will confirm if the PWM is being sent correctly from Pixhawk board to ESC. For this experiment you can deploy the the [example](https://in.mathworks.com/help/supportpkg/px4/ref/pwm-getting-started-px4.html) model. 
- Implement a direct pass through from RC receiver to PWM output in this [example](https://in.mathworks.com/help/supportpkg/px4/ref/pwm-getting-started-px4.html) model. Start from low throttle and see if motor spins correctly.
- Check if motors run as expected if you directly connect the RC receiver to ESC without Pixhawk.

**Reply by oh-hey-its-Jay** [reply]
Hello @vmanoj1996, 
The motors worked fine when running the "PX4 PWM" example model through connected IO.
When we input constant to the thrust roll pitch and yaw in our controller system model, all motors worked fine, when we connected our RC block the motors only stuttered/ twitched, but still responded to roll pitch and yaw commands. (also through connected IO) 
Currently using MATLAB R2022a

Thank you for the advice!

**Reply by seanatkins0** [reply]
Hi my name is Sean Atkins and I am on the same team as Jasmine. Here is the zipped Simulink file of the model we are using. We are using the monitor and tuned option via external mode when we are getting the "the base rate of the Simulink model 0 is...." error. [Flight_Test_Model_4_19_22.zip](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/files/8535416/Flight_Test_Model_4_19_22.zip)

**Reply by vmanoj1996** [reply]
Hey Sean,
3 things to look for,
- There are 1/s continuous time integrators in your model (Flight_Test_Model_4_19_22/Sensors). Replace them with discrete time integrators. 
- Sample time in PX4 uORB Read block is 1 ms. I am not sure if 1 ms rate is achievable on Pixhawk board. 
- manual_control_setpoint is an intermediate uorb topic published by another px4 module. Can you try using "Radio Control Transmitter" block instead. 
[Flight_Test_Model_4_19_22.zip](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/files/8537818/Flight_Test_Model_4_19_22.zip)

**Comment by patweatharva** [comment]
Hello all, 
I am Atharva from COEP Technological University majoring in Mechanical Engineering. I am really interested in working and collaborating on this project.

**Reply by vmanoj1996** [reply]
@patweatharva, Glad to see that you are interested. Fill out the form available at [this link](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Flight%20Controller%20Design%20and%20Hardware%20Deployment) in case you have not done it already. 

Feel free to post any questions below!

**Comment by ayushbhagat11** [comment]
![Screenshot (94)](https://github.com/mathworks/MATLAB-Simulink-Challenge-Project-Hub/assets/154872449/01551a39-f318-4f80-870c-dde0376e6e65)
hello, I am facing the above error, whenever I try to build the firmware in UAV toolbox support package for px4 autopilot. As Pixhawk 4 mini is not listed in the select controller, so I have chosen my board is not listed here. and choose the v5 c-make default file and facing the above error. please help in resolving the error.

**Reply by arunMathamkode** [reply]
Hi,
In your MATLAB command window, there will be build logs getting displayed. Can you please share that as a text file? That would be helpful in understanding what is going wrong. I would also recommend giving a try with px4_fmu-v5_default CMake and seeing if that is also causing a build failure.

**Comment by ayushbhagat11** [comment]
INTERNAL & PARTNERS

Hello,

We have tried with all the default cmake files present in the PX4 folder including px4_fmu-v5_default CMake, still causing the build failure in the end,  Also for the reference I have attached the log file, the error facing is regarding Nuttx.

With Regards,
Ayush Bhagat


5acXjzUk

**Reply by arunMathamkode** [reply]
Hi,
I could not find the complete build log. Can you please paste the entire build log that is appearing on the MATLAB command window (including the Nuttx error you mentioned) when you are building the Firmware? I would also encourage responding via the GitHub web page rather than replying via email. Replaying via email creates a new post and we loose the continuation of the discussion and not sure what happens to the email attachments.

**Reply by ayushbhagat11** [reply]
The custom uORB message "SL_CustomPX4uORB.msg" is successfully included in the PX4 firmware.
To use the message, set up PX4 firmware using the Hardware Setup screens and build the PX4 firmware for the selected CMake.
To log the message in ulog format, update the logger_topics.txt in SD Card.
SUBMODULEUPDATESTART
SUBMODULEUPDATEEND
-- PX4 version: v1.12.3
-- PX4 config file: /cygdrive/c/PX4/home/Firmware/boards/px4/fmu-v5/default.cmake
-- PX4 config: px4_fmu-v5_default
-- PX4 platform: nuttx
-- cmake build type: MinSizeRel
-- The CXX compiler identification is GNU 7.3.1
-- The C compiler identification is GNU 7.3.1
-- The ASM compiler identification is GNU
-- Found assembler: /cygdrive/c/PX4/toolchain/gcc-arm/bin/arm-none-eabi-gcc.exe
-- Check for working CXX compiler: /cygdrive/c/PX4/toolchain/gcc-arm/bin/arm-none-eabi-g++.exe
-- Check for working CXX compiler: /cygdrive/c/PX4/toolchain/gcc-arm/bin/arm-none-eabi-g++.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Detecting CXX compile features
-- Detecting CXX compile features - failed
-- Check for working C compiler: /cygdrive/c/PX4/toolchain/gcc-arm/bin/arm-none-eabi-gcc.exe
-- Check for working C compiler: /cygdrive/c/PX4/toolchain/gcc-arm/bin/arm-none-eabi-gcc.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Detecting C compile features
-- Detecting C compile features - failed
-- Enabling double FP precision hardware instructions
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.9", minimum required is "3") 
-- build type is MinSizeRel
-- PX4 ECL: Very lightweight Estimation & Control Library v1.9.0-rc1-591-gb3fed06
-- Building and including px4_io-v2_default
-- Found PythonInterp: /usr/bin/python3 (found version "3.6.9") 
-- Using C++11
-- Release build type: MinSizeRel
-- ROMFS: ROMFS/px4fmu_common
-- ROMFS:  Adding build/px4_fmu-v5_default/romfs_extras/px4_io-v2_default.bin -> /etc/extras/px4_io-v2_default.bin
-- ROMFS:  Adding boards/px4/fmu-v5/bootloader/px4_fmu-v5_bootloader.bin -> /etc/extras/bootloader.bin
-- ROMFS:  Adding platforms/nuttx/init/stm32f7/rc.board_arch_defaults -> /etc/init.d/rc.board_arch_defaults
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_defaults -> /etc/init.d/rc.board_defaults
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_sensors -> /etc/init.d/rc.board_sensors
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_extras -> /etc/init.d/rc.board_extras
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_mavlink -> /etc/init.d/rc.board_mavlink
-- Configuring done
-- Generating done
-- Build files have been written to: /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default
[0/1372] git submodule platforms/nuttx/NuttX/nuttx
[9/1372] Generating uORB topic headers
[10/1372] git submodule platforms/nuttx/NuttX/apps
[14/1372] Generating component_information/checksums.h
[15/1372] git submodule src/lib/ecl
[20/1372] Performing configure step for 'px4io_firmware'
FAILED: external/Stamp/px4io_firmware/px4io_firmware-configure 
cd /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default/external/Build/px4io_firmware && /usr/bin/cmake.exe -DCONFIG=px4_io-v2_default -GNinja /cygdrive/c/PX4/home/Firmware && /usr/bin/cmake.exe -E touch /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default/external/Stamp/px4io_firmware/px4io_firmware-configure
-- PX4 version: v1.12.3
CMake Error at cmake/px4_config.cmake:81 (message):
  PX4 config file not set, try one of
  airmind/mindpx-v2/default.cmake;ark/can-flow/canbootloader.cmake;ark/can-flow/debug.cmake;ark/can-flow/default.cmake;ark/can-gps/canbootloader.cmake;ark/can-gps/debug.cmake;ark/can-gps/default.cmake;atlflight/cmake_hexagon/bundle.cmake;atlflight/cmake_hexagon/fastrpc.cmake;atlflight/cmake_hexagon/hexagon_sdk.cmake;atlflight/cmake_hexagon/linux_app.cmake;atlflight/cmake_hexagon/qurt_flags.cmake;atlflight/cmake_hexagon/qurt_lib.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-arm-linux-gnueabihf.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-arm-oemllib32-linux-gnueabi.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-qurt.cmake;atlflight/eagle/cmake/upload.cmake;atlflight/eagle/default.cmake;atlflight/eagle/qurt.cmake;atlflight/excelsior/default.cmake;atlflight/excelsior/qurt.cmake;av/x-v1/default.cmake;beaglebone/blue/cmake/init.cmake;beaglebone/blue/cmake/upload.cmake;beaglebone/blue/default.cmake;bitcraze/crazyflie/default.cmake;bitcraze/crazyflie21/default.cmake;cuav/can-gps-v1/canbootloader.cmake;cuav/can-gps-v1/debug.cmake;cuav/can-gps-v1/default.cmake;cuav/nora/bootloader.cmake;cuav/nora/default.cmake;cuav/nora/test.cmake;cuav/x7pro/bootloader.cmake;cuav/x7pro/default.cmake;cuav/x7pro/test.cmake;cubepilot/cubeorange/bootloader.cmake;cubepilot/cubeorange/default.cmake;cubepilot/cubeorange/test.cmake;cubepilot/cubeyellow/default.cmake;cubepilot/cubeyellow/test.cmake;cubepilot/io-v2/default.cmake;emlid/navio2/cmake/upload.cmake;emlid/navio2/default.cmake;freefly/can-rtk-gps/canbootloader.cmake;freefly/can-rtk-gps/default.cmake;holybro/can-gps-v1/canbootloader.cmake;holybro/can-gps-v1/debug.cmake;holybro/can-gps-v1/default.cmake;holybro/durandal-v1/bootloader.cmake;holybro/durandal-v1/default.cmake;holybro/durandal-v1/test.cmake;holybro/kakutef7/default.cmake;holybro/pix32v5/default.cmake;holybro/pix32v5/test.cmake;modalai/fc-v1/default.cmake;modalai/fc-v1/rtps.cmake;modalai/fc-v1/test.cmake;modalai/fc-v2/bootloader.cmake;modalai/fc-v2/default.cmake;modalai/fc-v2/test.cmake;mro/ctrl-zero-f7-oem/default.cmake;mro/ctrl-zero-f7/default.cmake;mro/ctrl-zero-h7-oem/bootloader.cmake;mro/ctrl-zero-h7-oem/default.cmake;mro/ctrl-zero-h7/bootloader.cmake;mro/ctrl-zero-h7/default.cmake;mro/pixracerpro/bootloader.cmake;mro/pixracerpro/default.cmake;mro/x21-777/default.cmake;mro/x21/default.cmake;nxp/fmuk66-e/default.cmake;nxp/fmuk66-e/rtps.cmake;nxp/fmuk66-e/socketcan.cmake;nxp/fmuk66-v3/default.cmake;nxp/fmuk66-v3/rtps.cmake;nxp/fmuk66-v3/socketcan.cmake;nxp/fmuk66-v3/test.cmake;nxp/fmurt1062-v1/default.cmake;nxp/ucans32k146/canbootloader.cmake;nxp/ucans32k146/default.cmake;omnibus/f4sd/default.cmake;px4/fmu-v2/default.cmake;px4/fmu-v2/fixedwing.cmake;px4/fmu-v2/multicopter.cmake;px4/fmu-v2/rover.cmake;px4/fmu-v2/test.cmake;px4/fmu-v3/default.cmake;px4/fmu-v3/test.cmake;px4/fmu-v4/cannode.cmake;px4/fmu-v4/default.cmake;px4/fmu-v4/test.cmake;px4/fmu-v4pro/default.cmake;px4/fmu-v4pro/io-v2/default.cmake;px4/fmu-v4pro/test.cmake;px4/fmu-v5/ctrlalloc.cmake;px4/fmu-v5/debug.cmake;px4/fmu-v5/default.cmake;px4/fmu-v5/fixedwing.cmake;px4/fmu-v5/multicopter.cmake;px4/fmu-v5/optimized.cmake;px4/fmu-v5/rover.cmake;px4/fmu-v5/rtps.cmake;px4/fmu-v5/stackcheck.cmake;px4/fmu-v5/test.cmake;px4/fmu-v5/uavcanv0periph.cmake;px4/fmu-v5/uavcanv1.cmake;px4/fmu-v5x/base_phy_DP83848C.cmake;px4/fmu-v5x/default.cmake;px4/fmu-v5x/test.cmake;px4/fmu-v6x/bootloader.cmake;px4/fmu-v6x/default.cmake;px4/fmu-v6x/fixedwing.cmake;px4/fmu-v6x/multicopter.cmake;px4/fmu-v6x/test.cmake;px4/raspberrypi/cmake/upload.cmake;px4/raspberrypi/default.cmake;px4/sitl/ctrlalloc.cmake;px4/sitl/default.cmake;px4/sitl/nolockstep.cmake;px4/sitl/replay.cmake;px4/sitl/rtps.cmake;px4/sitl/test.cmake;scumaker/pilotpi/arm64.cmake;scumaker/pilotpi/cmake/upload.cmake;scumaker/pilotpi/default.cmake;spracing/h7extreme/default.cmake;uvify/core/default.cmake
Call Stack (most recent call first):
  CMakeLists.txt:142 (include)


-- Configuring incomplete, errors occurred!
[22/1372] Copying NuttX/apps to build/px4_fmu-v5_default/NuttX
[23/1372] Copying NuttX/nuttx to build/px4_fmu-v5_default/NuttX
ninja: build stopped: subcommand failed.



**This is the error I am facing everytime**

**Reply by arunMathamkode** [reply]
Are you working on windows or Linux? look like a toolchain issue.

**Reply by ayushbhagat11** [reply]
we are working on windows. what can be the possible steps to resolve this issue.

**Reply by arunMathamkode** [reply]
This looks like an issue with Cygwin toolchain. You can try uninstalling and reinstalling Cygwin and then clone a fresh PX4 Firmware ( you can refer [this page)](https://in.mathworks.com/help/supportpkg/px4/ug/setting-px4-toolchain-windows.html). Make sure the clone and update completed without errors. After this before building from the MATLAB Hardware setup screens, build the PX4 Firmware manually in Cygwin window using 'make px4_fmu-v5_default' command. See if you are facing any errors. Note that now you are building without MATLAB and Simulink in picture and if the build fails here, then it is due to some machine/environment specific issue and I am not really sure how to help you out here. You can report this to [PX4 issues.](https://github.com/PX4/PX4-Autopilot/issues) and you might get some help from PX4 community.

**Reply by ayushbhagat11** [reply]
Hello, thanks for your response and help. As per your suggestion, we tried uninstalling and reinstalling, cygwin and also cloned a fresh px4 firmware, it cloned and completed without any errors. Later we tried building the px4 firmware manually in cygwin window using 'make px4_fmu-v5_default' command, and we are facing a error there, can you please help out in resolving the error. Complete error is pasted below.

$ make px4_fmu-v5_default
[0/1] Re-running CMake...
-- PX4 version: v1.12.3
-- PX4 config file: /cygdrive/c/PX4/home/Firmware/boards/px4/fmu-v5/default.cmake
-- PX4 config: px4_fmu-v5_default
-- PX4 platform: nuttx
-- cmake build type: MinSizeRel
-- Enabling double FP precision hardware instructions
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.9", minimum required is "3")
-- build type is MinSizeRel
-- PX4 ECL: Very lightweight Estimation & Control Library v1.9.0-rc1-591-gb3fed06
-- Building and including px4_io-v2_default
-- Found PythonInterp: /usr/bin/python3 (found version "3.6.9")
-- Using C++11
-- Release build type: MinSizeRel
-- ROMFS: ROMFS/px4fmu_common
-- ROMFS:  Adding build/px4_fmu-v5_default/romfs_extras/px4_io-v2_default.bin -> /etc/extras/px4_io-v2_default.bin
-- ROMFS:  Adding boards/px4/fmu-v5/bootloader/px4_fmu-v5_bootloader.bin -> /etc/extras/bootloader.bin
-- ROMFS:  Adding platforms/nuttx/init/stm32f7/rc.board_arch_defaults -> /etc/init.d/rc.board_arch_defaults
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_defaults -> /etc/init.d/rc.board_defaults
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_sensors -> /etc/init.d/rc.board_sensors
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_extras -> /etc/init.d/rc.board_extras
-- ROMFS:  Adding boards/px4/fmu-v5/init/rc.board_mavlink -> /etc/init.d/rc.board_mavlink
-- Configuring done
-- Generating done
-- Build files have been written to: /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default
[0/1372] git submodule src/drivers/gps/devices
[2/1372] Performing configure step for 'px4io_firmware'
FAILED: external/Stamp/px4io_firmware/px4io_firmware-configure
cd /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default/external/Build/px4io_firmware && /usr/bin/cmake.exe -DCONFIG=px4_io-v2_default -GNinja /cygdrive/c/PX4/home/Firmware && /usr/bin/cmake.exe -E touch /cygdrive/c/PX4/home/Firmware/build/px4_fmu-v5_default/external/Stamp/px4io_firmware/px4io_firmware-configure
-- PX4 version: v1.12.3
CMake Error at cmake/px4_config.cmake:81 (message):
  PX4 config file not set, try one of
  airmind/mindpx-v2/default.cmake;ark/can-flow/canbootloader.cmake;ark/can-flow/debug.cmake;ark/can-flow/default.cmake;ark/can-gps/canbootloader.cmake;ark/can-gps/debug.cmake;ark/can-gps/default.cmake;atlflight/cmake_hexagon/bundle.cmake;atlflight/cmake_hexagon/fastrpc.cmake;atlflight/cmake_hexagon/hexagon_sdk.cmake;atlflight/cmake_hexagon/linux_app.cmake;atlflight/cmake_hexagon/qurt_flags.cmake;atlflight/cmake_hexagon/qurt_lib.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-arm-linux-gnueabihf.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-arm-oemllib32-linux-gnueabi.cmake;atlflight/cmake_hexagon/toolchain/Toolchain-qurt.cmake;atlflight/eagle/cmake/upload.cmake;atlflight/eagle/default.cmake;atlflight/eagle/qurt.cmake;atlflight/excelsior/default.cmake;atlflight/excelsior/qurt.cmake;av/x-v1/default.cmake;beaglebone/blue/cmake/init.cmake;beaglebone/blue/cmake/upload.cmake;beaglebone/blue/default.cmake;bitcraze/crazyflie/default.cmake;bitcraze/crazyflie21/default.cmake;cuav/can-gps-v1/canbootloader.cmake;cuav/can-gps-v1/debug.cmake;cuav/can-gps-v1/default.cmake;cuav/nora/bootloader.cmake;cuav/nora/default.cmake;cuav/nora/test.cmake;cuav/x7pro/bootloader.cmake;cuav/x7pro/default.cmake;cuav/x7pro/test.cmake;cubepilot/cubeorange/bootloader.cmake;cubepilot/cubeorange/default.cmake;cubepilot/cubeorange/test.cmake;cubepilot/cubeyellow/default.cmake;cubepilot/cubeyellow/test.cmake;cubepilot/io-v2/default.cmake;emlid/navio2/cmake/upload.cmake;emlid/navio2/default.cmake;freefly/can-rtk-gps/canbootloader.cmake;freefly/can-rtk-gps/default.cmake;holybro/can-gps-v1/canbootloader.cmake;holybro/can-gps-v1/debug.cmake;holybro/can-gps-v1/default.cmake;holybro/durandal-v1/bootloader.cmake;holybro/durandal-v1/default.cmake;holybro/durandal-v1/test.cmake;holybro/kakutef7/default.cmake;holybro/pix32v5/default.cmake;holybro/pix32v5/test.cmake;modalai/fc-v1/default.cmake;modalai/fc-v1/rtps.cmake;modalai/fc-v1/test.cmake;modalai/fc-v2/bootloader.cmake;modalai/fc-v2/default.cmake;modalai/fc-v2/test.cmake;mro/ctrl-zero-f7-oem/default.cmake;mro/ctrl-zero-f7/default.cmake;mro/ctrl-zero-h7-oem/bootloader.cmake;mro/ctrl-zero-h7-oem/default.cmake;mro/ctrl-zero-h7/bootloader.cmake;mro/ctrl-zero-h7/default.cmake;mro/pixracerpro/bootloader.cmake;mro/pixracerpro/default.cmake;mro/x21-777/default.cmake;mro/x21/default.cmake;nxp/fmuk66-e/default.cmake;nxp/fmuk66-e/rtps.cmake;nxp/fmuk66-e/socketcan.cmake;nxp/fmuk66-v3/default.cmake;nxp/fmuk66-v3/rtps.cmake;nxp/fmuk66-v3/socketcan.cmake;nxp/fmuk66-v3/test.cmake;nxp/fmurt1062-v1/default.cmake;nxp/ucans32k146/canbootloader.cmake;nxp/ucans32k146/default.cmake;omnibus/f4sd/default.cmake;px4/fmu-v2/default.cmake;px4/fmu-v2/fixedwing.cmake;px4/fmu-v2/multicopter.cmake;px4/fmu-v2/rover.cmake;px4/fmu-v2/test.cmake;px4/fmu-v3/default.cmake;px4/fmu-v3/test.cmake;px4/fmu-v4/cannode.cmake;px4/fmu-v4/default.cmake;px4/fmu-v4/test.cmake;px4/fmu-v4pro/default.cmake;px4/fmu-v4pro/io-v2/default.cmake;px4/fmu-v4pro/test.cmake;px4/fmu-v5/ctrlalloc.cmake;px4/fmu-v5/debug.cmake;px4/fmu-v5/default.cmake;px4/fmu-v5/fixedwing.cmake;px4/fmu-v5/multicopter.cmake;px4/fmu-v5/optimized.cmake;px4/fmu-v5/rover.cmake;px4/fmu-v5/rtps.cmake;px4/fmu-v5/stackcheck.cmake;px4/fmu-v5/test.cmake;px4/fmu-v5/uavcanv0periph.cmake;px4/fmu-v5/uavcanv1.cmake;px4/fmu-v5x/base_phy_DP83848C.cmake;px4/fmu-v5x/default.cmake;px4/fmu-v5x/test.cmake;px4/fmu-v6x/bootloader.cmake;px4/fmu-v6x/default.cmake;px4/fmu-v6x/fixedwing.cmake;px4/fmu-v6x/multicopter.cmake;px4/fmu-v6x/test.cmake;px4/raspberrypi/cmake/upload.cmake;px4/raspberrypi/default.cmake;px4/sitl/ctrlalloc.cmake;px4/sitl/default.cmake;px4/sitl/nolockstep.cmake;px4/sitl/replay.cmake;px4/sitl/rtps.cmake;px4/sitl/test.cmake;scumaker/pilotpi/arm64.cmake;scumaker/pilotpi/cmake/upload.cmake;scumaker/pilotpi/default.cmake;spracing/h7extreme/default.cmake;uvify/core/default.cmake
Call Stack (most recent call first):
  CMakeLists.txt:142 (include)


-- Configuring incomplete, errors occurred!
[6/1372] Generating git version header
ninja: build stopped: subcommand failed.
make: *** [Makefile:226: px4_fmu-v5_default] Error 1

**Reply by arunMathamkode** [reply]
Since the build is failing without MATLAB and Simulink in picture, it is due to some machine/environment specific issue and I am not really sure how to help you out here. You can report this to [PX4 issues.](https://github.com/PX4/PX4-Autopilot/issues) and you might get some help from PX4 community.

**Comment by ayushbhagat11** [comment]
![Screenshot (96)](https://github.com/mathworks/MATLAB-Simulink-Challenge-Project-Hub/assets/154872449/1193d3d0-6516-4637-93fa-5270b9df1e67)
Hello, I had a query, as the challenge states that "Pixhawk 4 mini flight controller" must be used to complete it, but its not listed in the dropdown list of px4 autopilot board in "UAV toolbox support package for px4 autopilot." Also can we use any of the controller mentioned like pixhawk cube orange inorder to complete the challenge. If yes, what all errors or difficulties we can face further. And if we have to go with only "Pixhawk 4 mini flight controller" then do we have to select "my board is not mentioned here" in the dropdown.

**Reply by arunMathamkode** [reply]
If you are using Pixhawk 4 mini you can choose 'My board not listed here' or ' PX4 Pixhawk 4' (Since both Pixhawk 4 and mini are fmu-v5). I recommend using 'PX4 Pixhawk 4' because you get some extra features like Connected I/O. You can also use Cube Orange also but make sure that you chose the physical airframe accordingly since their form factors are slightly different.  Here is the [link](https://in.mathworks.com/matlabcentral/answers/1868113-how-to-resolve-px4-firmware-build-failure-for-cube-orange-board) to some known build failure issues and their fixes for Cube Orange.

**Reply by ayushbhagat11** [reply]
Also is it necessary to use pixhawk 4 mini only, or we can use any flight controller, like pixhawk cube orange.

**Reply by arunMathamkode** [reply]
You can use any supported flight controller such as cube orange. Make sure that you chose the physical airframe accordingly.

**Comment by shiva-rajesh** [comment]
> Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project #217](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/edit/main/projects/Flight%20Controller%20Design%20and%20Hardware%20Deployment)

Hi, I am a final year student at Vellore Institute of Technology Bhopal and interested in working and collaborating on this project.

**Reply by ayushbhagat11** [reply]
hello, 
Good to see you are interested in the above challenge to complete, Please let me know how we can contribute more

**Reply by shiva-rajesh** [reply]
I would be grateful if I could get some guidance on where to start as this is the first time I am doing an industry level project with MATLAB Simulink.

**Reply by arunMathamkode** [reply]
Glad to see that you are interested in this project. You can find the instructions for setting up PX4 with Simulink [here](https://in.mathworks.com/help/uav/setup-and-configuration-px4.html). Additionally [these examples](https://in.mathworks.com/help/uav/px4-hitl.html) can help you get started.

Please feel free to post any questions or thoughts you may have below.

**Comment by shiva-rajesh** [comment]
Hi, I am Shiva Rajesh. I signed up for this challenge and set up the UAV toolbox support package for PX4 autopilots and tried to run the SITL simulation for a simple Simulink example model I found on MathWorks resource page. But whenever I run the model the jMAVsim simulation window does not pop up, instead i get the following error. I request your guidance to solve this.
![Screenshot (328)](https://github.com/user-attachments/assets/790c422c-6c50-4064-a889-f20f1c86fa79)

**Comment by AmitSilman** [comment]
https://radiolink.com.cn/doce/pro
![צילום מסך 2024-12-20 105001](https://github.com/user-attachments/assets/676f04dc-a0a1-4d78-8cc4-70ed7e04698d)
Hi all,
I want to start this project soon ,someone knows if this pixhawk controller is supported by simulink?

**Comment by Ajisegiri123** [comment]
Hello , I am an MSc student( Aerospace Engineering) at the University Of Salford. I had my bachelor degree in electrical and Electronics Engineering ..
Please I am interested in this project. How can I join and contribute to the project. I am new to Matlab design of  flight controller and simulink.

**Comment by AmitSilman** [comment]
Hi , we are working on it. You are more than welcome to join and work
together.
We can set an appointment if you'd like.
thanks,
amit
